File: Heap/1057.Campus_Bikes/Campus_Bikes.py
class Solution:
    def assignBikes(self, workers: List[List[int]], bikes: List[List[int]]) -> List[int]:
        q = []
        for i, w in enumerate(workers):
            for j, b in enumerate(bikes):
                d = abs(w[0] - b[0]) + abs(w[-1] - b[-1])
                q.append((d, i, j))
        q.sort()
        
        s = set() # assigned bike
        res = [-1] * len(workers) 
        for _, w_id, b_id in q:
            if len(s) == len(workers):
                    break
            if b_id not in s and res[w_id] == -1:
                s.add(b_id)
                res[w_id] = b_id
        return res

File: Heap/1057.Campus_Bikes/Campus_Bikes.java
class Solution {
    public int[] assignBikes(int[][] workers, int[][] bikes) {
        // index order
        PriorityQueue<int[]> q = new PriorityQueue<>((a, b) -> {
            int comp = Integer.compare(a[0], b[0]);
            if (comp == 0) {
                if (a[1] == b[1]) {
                    return Integer.compare(a[2], b[2]);
                }
                else {
                    return Integer.compare(a[1], b[1]);
                }
            }
            return comp;
        });
            
        for (int i = 0; i < workers.length; i++) {
            int[] worker = workers[i];
            for (int j = 0; j < bikes.length; j++) {
                int[] bike = bikes[j];
                int d = Math.abs(worker[0] - bike[0]) + Math.abs(worker[1] - bike[1]);
                q.add(new int[]{d, i, j});
            }
        }
        
        int[] ret = new int[workers.length];
        Arrays.fill(ret, -1);
        Set<Integer> set = new HashSet<>();
        while (set.size() < workers.length) {
            int[] tmp = q.poll();
            int wid = tmp[1], bid = tmp[2];
            if (ret[wid] == -1 && !set.contains(bid)) {
                ret[wid] = bid;
                set.add(bid);
            }
        }
        return ret;
    }
}